### PR TITLE
New source: HentaiVN

### PR DESCRIPTION
**Pull request by beerpiss**
- Source: HentaiVN (https://hentaivn.moe)
- Language: Vietnamese

It's a hentai source in Vietnamese. What more could you ask for?

### Notes
`.wireit` and `out` are .gitignored because I used [google/wireit](https://github.com/google/wireit) and aidoku-serve during development.

### CODE REVIEW

File: src/as/vi.hentaivn/src/parser.ts [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+  MangaStatus,
+  MangaViewer,
+  Page,
+  ValueRef,
+} from "aidoku-as/src/index";
+import { Constants } from "./constants";
+
+// MARK: Utilities
+function parseEmailProtected(data: string): string {
+  console.log(`[HentaiVN.Util.parseEmailProtected] Parsing ${data}`);
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:20]:
```suggestion
```

File: src/as/vi.hentaivn/src/parser.ts [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+  let email = "";
+  const key = parseInt(data.substr(0, 2), 16) as i64;
+  for (let n = 2; data.length - n; n += 2) {
+    const char = <i64> parseInt(data.substr(n, 2), 16) ^ key;
+    email += String.fromCharCode(<i32> char);
+  }
+  console.log(`[HentaiVN.Util.parseEmailProtected] Resolved to ${email}`);
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:27]:
```suggestion
```

File: src/as/vi.hentaivn/src/parser.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+}
+
+function parseAllEmailProtected(document: Html): Html {
+  let html = document.html();
+  const elements = document.select("span.__cf_email__").array();
+  console.log(
+    `[HentaiVN.Util.parseAllEmailProtected] Found ${elements.length} emails need parsing
+      ${html}
+    `,
+  );
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:38]:
```suggestion
```

File: src/as/vi.hentaivn/src/parser.ts [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+  for (let i = 0; i < elements.length; i++) {
+    const elem = elements[i];
+    const email = parseEmailProtected(elem.attr("data-cfemail"));
+
+    console.log(
+      `[HentaiVN.Util.parseAllEmailProtected] '${html}'.replace('${elem.outerHtml()}', '${email}')`,
+    );
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:45]:
```suggestion
```

File: src/as/vi.hentaivn/src/parser.ts [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+    html = html.replace(elem.outerHtml(), email);
+  }
+
+  console.log(`[HentaiVN.Util.parseAllEmailProtected] Done:
+    ${html}
+  `);
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:51]:
```suggestion
```

File: src/as/vi.hentaivn/src/parser.ts [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+  return urlSplit.join("/");
+}
+
+// MARK: Parser
+export class Parser {
+  parseNewOrCompletePage(document: Html): MangaPageResult {
+    const elements = document.select("li.item").array();
+    console.log(
+      `[HentaiVN.Parser.parseNewOrCompletePage] Found ${elements.length} mangas`,
+    );
```
Comment (Skittyblock) [inline src/as/vi.hentaivn/src/parser.ts:82]:
```suggestion
```

**Review by Skittyblock — CHANGES_REQUESTED** [review]
Also, the icon should be changed to fit iOS (I can do that if you need me to), and `package-lock.json` should be included since currently the `aidoku-as` package changes based on the github commit hash and not version number.

**Comment by beerpiss** [comment]
> Also, the icon should be changed to fit iOS (I can do that if you need me to), and `package-lock.json` should be included since currently the `aidoku-as` package changes based on the github commit hash and not version number.

That should be everything aside from the icon, can you help me out with that please?

**Comment by Skittyblock** [comment]
Yeah, I'll do it

**Pull request merged**